File: src/main/services/terminal/strategies/windowsTerminal.ts
import { spawn } from 'child_process'
import { exec } from 'child_process'
import { promisify } from 'util'
import type { TerminalStrategy, EnvObject } from './base'
import { escapeEnvValue, sanitizePath } from './escape'

const execAsync = promisify(exec)

export class WindowsTerminalStrategy implements TerminalStrategy {
  name = 'Windows Terminal'

  async isAvailable(): Promise<boolean> {
    if (process.platform !== 'win32') return false
    try {
      await execAsync('where wt.exe')
      return true
    } catch {
      return false
    }
  }

  async launch(path: string, env: EnvObject, cliCommand?: string): Promise<void> {
    const safePath = sanitizePath(path)
    const envSetCommands = Object.entries(env)
      .map(([key, value]) => `set "${key}=${escapeEnvValue(value)}"`)
      .join(' && ')

    // Use pushd to support UNC paths (e.g. \\psf\Home\...)
    let finalCommand = `pushd "${safePath}"`
    if (envSetCommands) {
      finalCommand += ` && ${envSetCommands}`
    }
    if (cliCommand) {
      finalCommand += ` && ${cliCommand}`
    }

    // Use spawn with args array to avoid nested quoting issues with wt.exe
    const child = spawn('wt.exe', ['-w', '0', 'new-tab', 'cmd', '/k', finalCommand], {
      detached: true,
      stdio: 'ignore',
    })
    child.unref()
  }
}

File: src-tauri/src/terminal/windows_terminal.rs
use super::{EnvObject, TerminalStrategy};
use std::process::Command;

pub struct WindowsTerminalStrategy;

impl TerminalStrategy for WindowsTerminalStrategy {
    fn name(&self) -> &str { "Windows Terminal" }

    fn is_available(&self) -> bool {
        #[cfg(target_os = "windows")]
        {
            Command::new("where")
                .arg("wt")
                .output()
                .map(|o| o.status.success())
                .unwrap_or(false)
        }
        #[cfg(not(target_os = "windows"))]
        { false }
    }

    fn launch(&self, path: &str, env: &EnvObject, cli_command: &str) -> Result<(), String> {
        let env_sets: String = env.iter()
            .map(|(k, v)| format!("$env:{}=\"{}\"", k, v))
            .collect::<Vec<_>>()
            .join("; ");

        let ps_command = if env_sets.is_empty() {
            format!("cd '{}'; {}", path, cli_command)
        } else {
            format!("cd '{}'; {}; {}", path, env_sets, cli_command)
        };

        Command::new("wt")
            .args(["new-tab", "powershell", "-NoExit", "-Command", &ps_command])
            .spawn()
            .map_err(|e| format!("Failed to launch Windows Terminal: {}", e))?;

        Ok(())
    }
}
